src/webkitgtk/web_context.rs

```diff
@@ -312,27 +313,58 @@ impl WebContextExt for super::WebContext {
   ) {
     let context = &self.os.context;
 
-    let download_started_handler = RefCell::new(download_started_handler);
+    let download_started_handler = Rc::new(RefCell::new(download_started_handler));
     let failed = Rc::new(RefCell::new(false));
 
     context.connect_download_started(move |_context, download| {
-      if let Some(uri) = download.request().and_then(|req| req.uri()) {
-        let uri = uri.to_string();
-        let mut download_location = download
-          .destination()
-          .map(PathBuf::from)
-          .unwrap_or_default();
+      let download_started_handler = download_started_handler.clone();
+      download.connect_decide_destination(move |download, suggested_filename| {
+        if let Some(uri) = download.request().and_then(|req| req.uri()) {
+          let uri = uri.to_string();
+
+          if let Some(download_started_handler) = download_started_handler.borrow_mut().as_mut() {
+            let mut download_location =
+              dirs::download_dir().unwrap_or_else(|| current_dir().unwrap_or_default());
+
+            let (mut suggested_filename, mut ext) = suggested_filename
+              .split_once('.')
+              .map(|(base, ext)| (base, format!(".{ext}")))
+              .unwrap_or((suggested_filename, "".to_string()));
+
+            // Weird bug but webkitgtk sets the extension to `.jpe` for JPEG files in these cases:
+            // <a href="data:image/jpeg;base64,..." download="" />
+            // <a href="/some-img.jpeg" download="some-name" />
+            //
+            // Other 4 character extensions like avif,webp seem to be set correctly.
+            if ext == ".jpe" {
+              ext = ".jpeg".to_string();
+            }
```

Comment:
I think we should keep `.jpe` because it looks like `.jpe` is a valid subset of the `.jpeg`... though it's being used rarely. 

This is kind of interesting so I googled the issue. The issue has existed for years that the system may use `.jpe` first than other aliases :\

https://bugs.kde.org/show_bug.cgi?id=439519
https://bugs.archlinux.org/task/54977